util/windows_logger.h

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2017 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+//
+// Logger implementation for Windows environments.
+
+#ifndef STORAGE_LEVELDB_UTIL_WINDOWS_LOGGER_H_
+#define STORAGE_LEVELDB_UTIL_WINDOWS_LOGGER_H_
+
+#include <array>
+#include <windows.h>
+#include "leveldb/env.h"
+
+#ifdef DeleteFile   // Win32 defines the DeleteFile macro
+#undef DeleteFile
+#endif
+
+namespace leveldb {
+
+class WindowsLogger : public Logger {
```

Comment:
This looks pretty similar to posix_logger.  I wonder if we can restructure things to share that code.  That code already has removed the gettid dependency by taking a function pointer.  We could do a similar thing for getting the current local time into a string.

Reply:
I'll work on this as well, and edit my commit history to reflect this. Thanks.

Reply:
@ghemawat I've refactored them, changed `PosixLogger` to be named `FileLogger`, and added a method called `AppendDateTo` which takes an iterator range for the first and last character in the buffer. This function differs on Windows and POSIX, but allows the rest of the code to be shared between POSIX and Windows.

**Windows**

```cpp
char* FileLogger::AppendDateTo(char* first, char* last)
{
  const uint64_t thread_id = (*gettid_)();

  SYSTEMTIME st;
  GetLocalTime(&st);
  first += snprintf(first, last - first,
                "%04d/%02d/%02d-%02d:%02d:%02d.%03d %llx ",
                st.wYear,
                st.wMonth,
                st.wDay,
                st.wHour,
                st.wMinute,
                st.wSecond,
                st.wMilliseconds,  // No microsecond precision, use milliseconds
                static_cast<long long unsigned int>(thread_id));

  return first;
}
```

**POSIX**

```cpp
char* FileLogger::AppendDateTo(char* first, char* last)
{
  const uint64_t thread_id = (*gettid_)();

  struct timeval now_tv;
  gettimeofday(&now_tv, NULL);
  const time_t seconds = now_tv.tv_sec;
  struct tm t;
  localtime_r(&seconds, &t);

  first += snprintf(first, last - first,
                "%04d/%02d/%02d-%02d:%02d:%02d.%06d %llx ",
                t.tm_year + 1900,
                t.tm_mon + 1,
                t.tm_mday,
                t.tm_hour,
                t.tm_min,
                t.tm_sec,
                static_cast<int>(now_tv.tv_usec),
                static_cast<long long unsigned int>(thread_id));

  return first;
}
```